pkg/overlay/overlay.go

```diff
@@ -154,41 +188,21 @@ func MountWithOptions(contentDir, source, dest string, opts *Options) (mount spe
 		overlayOptions = fmt.Sprintf("lowerdir=%s,upperdir=%s,workdir=%s,private", escapeColon(source), upperDir, workDir)
 	}
 
-	if unshare.IsRootless() {
-		mountProgram := ""
-
-		mountMap := map[string]bool{
-			".mount_program":         true,
-			"overlay.mount_program":  true,
-			"overlay2.mount_program": true,
-		}
-
-		for _, i := range opts.GraphOpts {
-			s := strings.SplitN(i, "=", 2)
-			if len(s) != 2 {
-				continue
-			}
-			key := s[0]
-			val := s[1]
-			if mountMap[key] {
-				mountProgram = val
-				break
-			}
+	mountProgram := findMountProgram(opts.GraphOpts)
+	if mountProgram != "" {
+		if err := mountWithMountProgram(mountProgram, overlayOptions, mergeDir); err != nil {
+			return mount, err
 		}
-		if mountProgram != "" {
-			cmd := exec.Command(mountProgram, "-o", overlayOptions, mergeDir)
 
-			if err := cmd.Run(); err != nil {
-				return mount, errors.Wrapf(err, "exec %s", mountProgram)
-			}
+		mount.Source = mergeDir
+		mount.Destination = dest
+		mount.Type = "bind"
+		mount.Options = []string{"bind", "slave"}
+		return mount, nil
+	}
 
-			mount.Source = mergeDir
-			mount.Destination = dest
-			mount.Type = "bind"
-			mount.Options = []string{"bind", "slave"}
-			return mount, nil
-		}
-		/* If a mount_program is not specified, fallback to try mount native overlay.  */
+	if unshare.IsRootless() {
+		/* If a mount_program is not specified, fallback to try mounting native overlay.  */
```

Comment:
One doubt: We used to apply user's extended attributes for both `root` and `rootless` when we were falling back to `native overlay`. Does this check introduces a behavior change, I thought this is a `no-op` for root users.

Reply:
previously it was also only set for rootless.

userxattr is useful only for rootless that cannot use the privileged overlay xattr

Reply:
ah i see got it. My bad.